staking-contract/contracts/StakingRewardSystem.sol

```diff
@@ -0,0 +1,221 @@
+    function _addRewardToken(address _rewardToken) internal {
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        rewardTokens.push(_rewardToken);
+        isRewardToken[_rewardToken] = true;
+        emit RewardTokenAdded(_rewardToken, msg.sender);
+    }
+
+    // Function to set the minimum staking period
+    function setMinStakingPeriod(uint256 _period) external onlyOwner {
+        if (_period <= 0) revert InvalidInput("Minimum staking period must be greater than zero");
+        uint256 prevminStakingPeriod=minStakingPeriod;
+        minStakingPeriod = _period*86400;
+        emit MinStakingPeriodUpdated(prevminStakingPeriod,minStakingPeriod, msg.sender);
+    }
+
+    // Function to set the early withdrawal penalty percentage
+    function setEarlyWithdrawalPenalty(uint256 _penalty) external onlyOwner {
+        if (_penalty <= 0 || _penalty > 1e18) revert InvalidInput("Penalty must be between 0 and 100%");
+        uint256 prevPenalty=earlyWithdrawalPenalty;
+        earlyWithdrawalPenalty = _penalty;
+        emit EarlyWithdrawalPenaltyUpdated(prevPenalty, earlyWithdrawalPenalty, msg.sender);
+    }
+
+    // Function to set the reward rate
+    function setRewardRate(uint256 _rate) external onlyOwner {
+        if (_rate <= 0) revert InvalidInput("Reward rate must be greater than zero");
+        uint256 preRewardRate=rewardRate;
+        rewardRate = _rate;
+        emit RewardRateUpdated(preRewardRate, rewardRate, msg.sender);
+    }
+
+    // Function to get the list of reward tokens
+    function getRewardTokens() external view returns (address[] memory) {
+        return rewardTokens;
+    }
+
+    // Function to get the staked amount of a user for a specific reward token
+    function getStakedAmount(address _user, address _rewardToken) external view returns (uint256) {
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (!isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        Stake storage userStake = userStakes[_user][_rewardToken];
+        if (userStake.amount == 0) revert UserNotExist();
+        return userStakes[_user][_rewardToken].amount;
+    }
+
+    // Function to get all staking data for a user, including the amount staked and the rewards earned
+    function getUserStakeData(address _user) external view returns (Stake[] memory) {
+        uint256 tokenCount = rewardTokens.length;
+        Stake[] memory stakesData = new Stake[](tokenCount);
+        bool userExists = false;
+        for (uint256 i = 0; i < tokenCount; i++) {
+            address rewardToken = rewardTokens[i];
+            Stake storage userStake = userStakes[_user][rewardToken];
+            if (userStake.amount > 0) {
+                userExists = true;
+            }
+            stakesData[i] = Stake({
```

Comment:
Don't you think you should simply push instead of assigning at each position?
Because suppose if there are 1000 tokens and user only staked on the 999th position and if we have an api to fetch user staked data and it has pagination of 100 records per page.
Then from this function we will have to move till last page to see the user staked amount?